cpmpy/expressions/core.py

```diff
@@ -137,12 +159,44 @@ def __repr__(self):
     def __hash__(self):
         return hash(self.__repr__())
 
+    def has_subexpr(self):
+        """ Does it contains nested Expressions (anything other than a _NumVarImpl or a constant)?
+            Is of importance when deciding whether certain transformations are needed
+            along particular paths of the expression tree.
+            Results are cached for future calls and reset when the expression changes
+            (in-place argument update).
+        """
+
+        def recursive_has_subexpr(lst) -> bool:
+            """ Recursive implementation to handle nested lists of expressions.
+            """
+            for el in lst:
+                if isinstance(el, Expression):
+                    if not el.is_leaf():  # NDVarArrays are Expr and any_list, so they are covered too (allows early-exit for decision var arrays)
+                        return True
+                elif is_any_list(el) and recursive_has_subexpr(el): # recursively call on list-like
+                    return True
+            return False
+
+        # If not an Expression (e.g. list-like) or _has_subexpr has not been computed before / has been reset
+        if not hasattr(self, '_has_subexpr'):
+            # args can have lists of lists... -> need for recursive implementation
+            self._has_subexpr = recursive_has_subexpr(self.args)
+
+        return self._has_subexpr
+
     def is_bool(self):
         """ is it a Boolean (return type) Operator?
             Default: yes
         """
         return True
 
+    def is_leaf(self):
```

Comment:
since we have `has_subexpr()` maybe we should name it `is_leafexpr()` because 'leaf' as a word is not very descriptive

Reply:
actually, what `Expressions` are leafs? Only variables? so should be `is_var()`?

Reply:
or a lists of vars/constants yes